### DISCUSSION TITLE
Suspense - hydration error on page refresh

### DISCUSSION THREAD

**Discussion by nemanjam** [vercel/next.js#36582] [Help]
I have Next.js server side rendered app and I struggle with hydration and unmatched client-server state. It happens in both dev and prod build. Here are screenshots of errors:

![image](https://user-images.githubusercontent.com/9990165/166091619-e089383a-fb88-45a5-8a8c-d83c907fa7d1.png)
![image](https://user-images.githubusercontent.com/9990165/166091624-990591f5-ae9d-46e8-ba03-58e4b240b38c.png)

I pass me user (authed user) and page data through getServerSideProps and ReactQuery queries. It is done in two levels, me is fetched in MeProvider and page data inside a page component (see code snippets).

I use React 18. I can solve this error by preventing rendering unmounted app with `{isMounted ? children : null}` (see MeProvider code) but then I have ugly white flashing of all content including Navbar that disappears from screen on page navigation.

I have one Suspense component in the app root, and another Suspense in PageLayout that wraps page components.

Any idea how can I solve this? Eliminate error without blocking the render.

**/pages/\_app.tsx**
```tsx
  const App = ({
    Component,
    pageProps: { session, dehydratedState, ...pageProps },
  }: AppProps) => {
    return (
      <SuspenseWrapper errorFallbackType="screen" loaderType="screen">
        {/* ...other providers */}
        <QueryClientProvider client={queryClient}>
          <Hydrate state={dehydratedState}>
            <Component {...pageProps} />
          </Hydrate>
          <ReactQueryDevtools />
        </QueryClientProvider>
      </SuspenseWrapper>
    );
  };
```
**/pages/index.tsx**
```tsx
  const Home: FC = () => {
    return (
      <PageLayout>
        <HomeView /> // this fetches data with React Query queries
      </PageLayout>
    );
  };
  
  export const getServerSideProps: GetServerSideProps = async ({ req, res }) => {
    // ...
  
    // prefetch page props and me for MeProvider (auth user)
    const queryClient = new QueryClient();
    await queryClient.prefetchQuery([QueryKeys.POSTS_HOME, 1], () => posts);
    await queryClient.prefetchQuery(filterEmptyKeys([QueryKeys.ME, me?.id]), () => me);
  
    return {
      props: {
        dehydratedState: dehydrate(queryClient),
      },
    };
  };
```
**/layouts/PageLayout.tsx**
```tsx
  const PageLayout: FC<Props> = ({ children, noPaddingTop }) => {
    const b = withBem('page-layout');
  
    return (
      <MeProvider> // MeProvider does fetching with useMe() for auth user
        <div className={b()}>
          <Navbar />
          <div className={b('navbar-placeholder')} />
  
          <main>
            <SuspenseWrapper errorFallbackType="page" loaderType="page">
              {children} // pages fetch their data
            </SuspenseWrapper>
          </main>
  
          <Footer />
        </div>
      </MeProvider>
    );
  };
```
**/providers/Me.tsx**
```tsx
  const MeProvider: FC<ProviderProps> = ({ children }) => {
    const isMounted = useIsMounted();
  
    const { data } = useMe();
  
    // hydration error on page refresh
    return <MeContext.Provider value={{ me: data }}>{children}</MeContext.Provider>;
  
    // this fixes hydration error but causes white flashing (rerender) when page is changed
    return (
      <MeContext.Provider value={{ me: data }}>
        {isMounted ? children : null}
      </MeContext.Provider>
    );
  };
```

**Comment by krazyjakee** [comment]
What a nightmare...
Have they deprecated `getServerSideProps` and not written it anywhere? Surely this page: https://nextjs.org/docs/advanced-features/react-18/overview should contain a migration guide if we're supposed to code things differently?

**Reply by icyJoseph** [reply]
What are you on about? `getServerSideProps` is still as valid as it was before.

Unfortunately there has been some strange communication regarding Suspense, and data fetching. There's various ongoing discussion on how it should work and how it should be handled by data fetching libraries. As far as I know, nobody has told developers to go out and change their code to use Suspense boundaries for data fetching, at day zero of React 18.

**Reply by 0xD4V1NC1** [reply]
I had a similar issue with using Suspense and data fetching with SWR with NextJS. For me what solved was adding state to the component that did the data fetching... i.e. 

```
import React, {useState, useEffect} from 'react'

const FetchedData = ()  =>  {
  const [hasMounted, setHasMounted] = useState(false);
  const {data} = useSWR(hasMounted ? 'someApiUrl': null, fetcher, { suspense: true })
  
  useEffect( () => setHasMounted(true), []);
  return <p>{data}</p>;
}
```
And in the parent component using Suspense...

```
const HomePage = () => {
  return(
      <div>
        Some content
        <Suspense fallback={<p>is loading... </p>} >
          <FetchedData />
        </Suspense>
      </div>
    );
    };
```

**Reply by 0xD4V1NC1** [reply]
Based off of what I found from this article recommend by NextJS when dealing with Hydration issues https://www.joshwcomeau.com/react/the-perils-of-rehydration/

**Reply by 0xD4V1NC1** [reply]
If that doesn't help you, I hope it helps someone else lol...

**Reply by nemanjam** [reply]
I think I did same with `enabled` prop in React Query and it didn't help.
